field/FieldPointerUtilities.F90

```diff
@@ -846,40 +848,78 @@ subroutine MAPL_FieldGetLocalElementCount(field,local_count,rc)
 
       call ESMF_FieldGet(field,rank=rank,typekind=tk,_RC)
       if (tk == ESMF_TypeKind_R4) then
-         if (rank==1) then
+         select case(rank)
+         case(1)
             call ESMF_FieldGet(field,0,farrayptr=r4_1d,_RC)
             local_count = shape(r4_1d)
-         else if (rank ==2) then
+         case(2)
             call ESMF_FieldGet(field,0,farrayptr=r4_2d,_RC)
             local_count = shape(r4_2d)
-         else if (rank ==3) then
+         case(3)
             call ESMF_FieldGet(field,0,farrayptr=r4_3d,_RC)
             local_count = shape(r4_3d)
-         else if (rank ==4) then
+         case(4)
             call ESMF_FieldGet(field,0,farrayptr=r4_4d,_RC)
             local_count = shape(r4_4d)
-         else
+         case default
             _FAIL("Unsupported rank")
-         end if
+         end select
       else if (tk == ESMF_TypeKind_R8) then
-         if (rank==1) then
+         select case(rank)
+         case(1)
             call ESMF_FieldGet(field,0,farrayptr=r8_1d,_RC)
             local_count = shape(r8_1d)
-         else if (rank ==2) then
+         case(2)
             call ESMF_FieldGet(field,0,farrayptr=r8_2d,_RC)
             local_count = shape(r8_2d)
-         else if (rank ==3) then
+         case(3)
             call ESMF_FieldGet(field,0,farrayptr=r8_3d,_RC)
             local_count = shape(r8_3d)
-         else if (rank ==4) then
+         case(4)
             call ESMF_FieldGet(field,0,farrayptr=r8_4d,_RC)
             local_count = shape(r8_4d)
-         else
+         case default
             _FAIL("Unsupported rank")
-         end if
+         end select
+      else if (tk == ESMF_TypeKind_I4) then
+         select case(rank)
+         case(1)
+            call ESMF_FieldGet(field,0,farrayptr=i4_1d,_RC)
+            local_count = shape(i4_1d)
+         case(2)
+            call ESMF_FieldGet(field,0,farrayptr=i4_2d,_RC)
+            local_count = shape(i4_2d)
+         case(3)
+            call ESMF_FieldGet(field,0,farrayptr=i4_3d,_RC)
+            local_count = shape(i4_3d)
+         case(4)
+            call ESMF_FieldGet(field,0,farrayptr=i4_4d,_RC)
+            local_count = shape(i4_4d)
+         case default
+            _FAIL("Unsupported rank")
+         end select
+      else if (tk == ESMF_TypeKind_I8) then
+         select case(rank)
+         case(1)
+            call ESMF_FieldGet(field,0,farrayptr=i8_1d,_RC)
+            local_count = shape(i8_1d)
+         case(2)
+            call ESMF_FieldGet(field,0,farrayptr=i8_2d,_RC)
+            local_count = shape(i8_2d)
+         case(3)
+            call ESMF_FieldGet(field,0,farrayptr=i8_3d,_RC)
+            local_count = shape(i8_3d)
+         case(4)
+            call ESMF_FieldGet(field,0,farrayptr=i8_4d,_RC)
+            local_count = shape(i8_4d)
+         case default
+            _FAIL("Unsupported rank")
+         end select
```

Comment:
You still need a final `_RETURN(_SUCCESS)` at the end of the procedure.  (Can't make a comment down there or I'd fix it for you.)  The reason is just in case anyone ever adds some logic that might skip the above.  If procedure was shorter I would not worry.

Reply:
OK. I'll add a final _RETURN(_SUCCESS).